### PR TITLE
[cmake] Internal depends download macro refactor V2

### PR DESCRIPTION
**Pull request by fuzzard**
## Description
This takes #20905 and then takes it to the conclusion i had intended.

Internal depends builds are refactored out into macros/functions to more easily template an internal dep build

## Motivation and context
Had phone/internet service out for the day, so got bored and spent some time working through this.
This would potentially be a way to migrate to cmake from tools/depends. Atm, the non cmake targets (DAV1D, FSTRCMP, UDFREAD) dont build outside of host=target, but the cmake targets can at least be built on osx for my testing purposes.

~~As always FFMPEG is untouched. If #21047 gets merged, i'll revisit these changes on it.~~
~~gtest is also not the full set of changes. Wasnt a priority, and can be done later for uniformity.~~

Ffmpeg and gtest are done now as well.

Theres a lot of commits, however i can squash a number of them into similar changes if desired.
This could supersede #20905 if desired to go all the way.

I'm ok if this gets rejected, i had a lot of time to kill today, haha. Always happy for any feedback, or opinions on changes.

## How has this been tested?
configure and build of targets where possible on osx.
minimal testing for those targets that dont build on osx on a ubuntu vm

## What is the effect on users?
n/a

## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

**Comment by fuzzard** [comment]
jenkins build this please

File: cmake/scripts/common/ModuleHelpers.cmake [hunk 1]
```diff
@@ -66,4 +72,75 @@ macro(SETUP_BUILD_VARS)
...
+
+  if(BUILD_IN_SOURCE)
+    set(BUILD_IN_SOURCE BUILD_IN_SOURCE ${BUILD_IN_SOURCE})
+  endif()
+
+  if(NOT BUILD_BYPRODUCTS)
+    set(BUILD_BYPRODUCTS ${CMAKE_BINARY_DIR}/${CORE_BUILD_DIR}/lib/${${MODULE}_BYPRODUCT})
+  else()
+    set(BUILD_BYPRODUCTS ${BUILD_BYPRODUCTS})
+  endif()
```
Comment (wsnipex) [inline cmake/scripts/common/ModuleHelpers.cmake:129]: you could invert the if for consistency and readablility

**Review by wsnipex — APPROVED** [review]

File: cmake/scripts/common/ModuleHelpers.cmake (hunk 1 above)
Comment (fuzzard) [reply cmake/scripts/common/ModuleHelpers.cmake:129]: Will do

Comment (fuzzard) [reply cmake/scripts/common/ModuleHelpers.cmake:133]: Have fixed that.
Also changed to BUILD_BYPRODUCTS is an optional. In the case of ffmpeg, we dont set a byproduct (it has more than one essentially), so checks have been put in to handle no byproduct, and to not set the byproduct vars if they arent found in the VERSION file.
This is also the case for MODULE_LIBRARY, if there isnt a byproduct, dont generate that LIBRARY var automatically. This just means the dev has to set the MODULE_LIBRARY var if its required. in the case of ffmpeg, we dont use FFMPEG_LIBRARY, so no need for it to be manually set.

**Review by wsnipex — APPROVED** [review]

**Pull request merged**